content/docs/learn/typed-errors/validation.md

```diff
@@ -0,0 +1,323 @@
+              .bind()
+          }
+          ensureNotNull(validatedAuthors.toNonEmptyListOrNull()) { NoAuthors }
+        }
+      ) { _, authorsNel ->
+        Book(title, authorsNel)
+      }
+    }
+  }
+}
+```
+<!--- KNIT example-validation-07.kt -->
+
+This additional check is quite complex, so let's unravel it step by step:
+- We use `authors.withIndex()` to create an iterable containing the values
+  in conjunction with the index they live in, this is necessary to create
+  the right `EmptyAuthor` error value.
+- With `mapOrAccumulate` we state that we want to perform some validation over
+  a collection of elements, accumulating each possible error.
+- The call to `Author(it.value)` return an `Either` with the wrong error type
+  (`EmptyAuthorName` instead of `EmptyAuthor`). To transform this value we
+  use the `recover` extension function.
+
+:::tip Recover vs. Map Left
+
+Another possibility would be to use `mapLeft { EmptyAuthor(it.index) }`.
+The difference between `recover` and `mapLeft` is that the latter only transforms
+the error value, whereas in the former you can use any typed error computation.
+
+:::
+
+- Finally, we use `.bind()` to embed the `Either` into the computation block.
+  Essentially, every time you are using a value of type `Either` inside an
+  `either` (or any other `Raise` block), such a call is required.
+
+The result of the mapping is a `List<Author>`, that we can now use to create the
+final `Book`. This value is available in the last lambda of `zipOrAccumulate`,
+that we've called `validatedAuthors` in the code above.
+
+
+<!--- INCLUDE
+sealed interface BookValidationError
+object EmptyTitle: BookValidationError
+object NoAuthors: BookValidationError
+data class EmptyAuthor(val index: Int): BookValidationError
+
+object EmptyAuthorName
+
+data class Author private constructor(val name: String) {
+  companion object {
+    operator fun invoke(name: String): Either<EmptyAuthorName, Author> = either {
+      ensure(name.isNotEmpty()) { EmptyAuthorName }
+      Author(name)
+    }
+  }
+}
+
+data class Book private constructor(val title: String, val authors: NonEmptyList<Author>) {
+  companion object {
+    operator fun invoke(
+      title: String, authors: Iterable<String>
+    ): Either<NonEmptyList<BookValidationError>, Book> = either {
+      zipOrAccumulate(
+        { ensure(title.isNotEmpty()) { EmptyTitle } },
+        { 
+          val validatedAuthors = authors.withIndex().map { a ->
+            Author(a.value).mapLeft { EmptyAuthor(a.index) }
+          }.bindAll()
+          ensureNotNull(validatedAuthors.toNonEmptyListOrNull()) { NoAuthors }
+        }
+      ) { _, validatedAuthors ->
+        Book(title, validatedAuthors)
+      }
+    }
+  }
+}
+-->
+<!--- KNIT example-validation-08.kt -->
+
+:::info Bind All
+
+Another way to write the code above is creating a list of `Either` using
+`mapOrAccumulate`, and then using `.bindAll()` at the very end.
```

Comment:
You mention `mapOrAccumulate` in the text. However, the code example uses just `map`. Both cases compile because `zipOrAccumulate` already provides the `RaiseAccumulate` context, but I think we should explain it or fix the text.

Reply:
I've updated and clarified the text, does it look better now?

Reply:
Looks perfect now, thanks @serras